File: Java/LCP_07-1.java
class Solution {
    private Map<Integer, List<Integer>> map;
    private int count;
    
    private void visit(int r0, int k, int n) {
        if (k == 0) {
            if (r0 == n - 1) {
                ++count;
            }
            return;
        }
        if (map.containsKey(r0)) {
            for (int i : map.get(r0)) {
                visit(i, k - 1, n);
            }
        }
        else {
            return;
        }
    }
    
    public int numWays(int n, int[][] relation, int k) {
        map = new HashMap<>();
        for (int[] r : relation) {
            List<Integer> lst = map.getOrDefault(r[0], new LinkedList<Integer>());
            lst.add(r[1]);
            map.put(r[0], lst);
        }
        count = 0;
        visit(0, k, n);
        return count;
    }
}

File: python/LCP_07-1.py
class Solution:
    def numWays(self, n: int, relation: List[List[int]], k: int) -> int:
        id2list = {}
        for r in relation:
            if r[0] in id2list:
                id2list[r[0]].append(r[1])
            else:
                id2list[r[0]] = [r[1]]
        
        self.count = 0
        def dfs(n, k, idx):
            if k == 0:
                if idx == n - 1:
                    self.count += 1
                return
            if idx in id2list:
                for i in id2list[idx]:
                    dfs(n, k - 1, i)
        
        dfs(n, k, 0)
        return self.count
